irc/src/connection.rs

```diff
@@ -26,16 +27,45 @@ pub enum Security<'a> {
     },
 }
 
+#[derive(Debug, Clone)]
+pub enum Proxy {
+    Socks5 {
+        host: String,
+        port: u16,
+        username: String,
+        password: String,
+    },
+}
+
 #[derive(Debug, Clone)]
 pub struct Config<'a> {
     pub server: &'a str,
     pub port: u16,
     pub security: Security<'a>,
+    pub proxy: Option<Proxy>,
 }
 
 impl<Codec> Connection<Codec> {
     pub async fn new(config: Config<'_>, codec: Codec) -> Result<Self, Error> {
-        let tcp = TcpStream::connect((config.server, config.port)).await?;
+        let target = (config.server, config.port);
+        let tcp = match config.proxy {
+            None => TcpStream::connect(target).await?,
+            Some(Proxy::Socks5 {
+                host,
+                port,
+                username,
+                password,
+            }) => {
+                let proxy = (host.as_str(), port);
+                if username.trim().is_empty() {
+                    Socks5Stream::connect(proxy, target).await?.into_inner()
+                } else {
+                    Socks5Stream::connect_with_password(proxy, target, &username, &password)
+                        .await?
+                        .into_inner()
```

Comment:
Does calling into_inner eliminate the proxy? We may need to use an enum to keep the proxy type for it to work

Reply:
as I tested, the `into_inner` provides the `TcpStream` chained with the socks5 proxy.